### PR TITLE
[COST-3726] Node Capacity

### PR DESCRIPTION
**Pull request by myersCody**
## Jira Ticket

[COST-3726](https://issues.redhat.com/browse/COST-3726)

Just checking the logic before I apply it anywhere else.

## Description

This change will calculate the node capacity and the unused portions of the request & capacity.

## Testing

compute:
```
},
                            "request": {
                                "value": 120.0,
                                "units": "Core-Hours",
                                "unused": 58.29158,
                                "unused_percent": 48.57631666666666
                            },
                            "limit": {
                                "value": 120.0,
                                "units": "Core-Hours"
                            },
                            "capacity": {
                                "value": 144.0,
                                "units": "Core-Hours",
                                "unused": 24.0,
                                "unused_percent": 16.666666666666668,
                                "instances": 6.0,
                                "instance_type": "Core"
                            },
```
Memory:
```
"request": {
                                "value": 240.0,
                                "units": "GB-Hours",
                                "unused": 0,
                                "unused_percent": 0.0
                            },
                            "limit": {
                                "value": 480.0,
                                "units": "GB-Hours"
                            },
                            "capacity": {
                                "value": 384.0,
                                "units": "GB-Hours",
                                "unused": 96.74,
                                "unused_percent": 25.192708333333332,
                                "instances": 16.0,
                                "instance_type": "GB"
                            },
```
Storage:
```
"request": {
                                "value": 1.290322580645162,
                                "units": "GB-Mo",
                                "unused": 0.625268817204302,
                                "unused_percent": 48.45833333333338
                            },
                            "capacity": {
                                "value": 1.290322580645162,
                                "units": "GB-Mo",
                                "unused": 0.0,
                                "unused_percent": 0.0,
                                "instances": 40.0,
                                "instance_type": "GB"
                            },
```

## Notes

...

### CODE REVIEW

File: koku/api/report/ocp/provider_map.py [hunk 1]
```diff
@@ -464,6 +472,9 @@ def __init__(self, provider, report_type):
                             "capacity": Sum("persistentvolumeclaim_capacity_gigabyte_months"),
                         },
                         "default_ordering": {"usage": "desc"},
+                        "capacity_aggregate": {
+                            "node": {"capacity": Sum("persistentvolumeclaim_capacity_gigabyte_months")},  # We sum
```
Comment (myersCody) [inline koku/api/report/ocp/provider_map.py:476]: aah such elegant comments from me (will clean up)

File: koku/api/report/ocp/query_handler.py [hunk 2]
```diff
@@ -31,6 +34,51 @@
...
+            self.by_level[level_value] += cap_value
+            self.resolution_total[month] += cap_value
+            self.resolution_level_total[month][level_value] += cap_value
+
+
+def _calculate_unused(row):
+    """Calculates the unused portions of the capacity & request."""
+    # Populate unused request and capacity
+    capacity = row.get("capacity", Decimal(0))
+    if capacity == 0:
```
Comment (lcouzens) [inline koku/api/report/ocp/query_handler.py:66]: Optional: Is it worth guarding against negative values here too? and on line 77 below. I know its not really likely but if we did get a negative value from a custom report that could be prevented here.

Comment (myersCody) [inline koku/api/report/ocp/query_handler.py:66]:
```suggestion
    if capacity <= 0:
```

**Review by lcouzens — APPROVED** [review]

**Pull request merged**